### ISSUE TITLE
Metal buffer pool grows beyond physical RAM on small unified-memory devices

### ISSUE THREAD

**Issue by cjchanh** [huggingface/candle#3464]
## Summary
On an 8 GB Apple Silicon machine, I observed Candle's Metal buffer pool grow well past physical RAM during a single decode request, which appears to trigger severe unified-memory swap thrash.

This was instrumented while integrating a fused 4-bit Metal path in Cake, but the behavior I am reporting here is specifically about Candle's Metal allocator / pooled buffer lifetime.

## Environment
- `candle-core = 0.9.2`
- `candle-metal-kernels = 0.9.2`
- hardware: MacBook Air M1, 8 GB unified memory
- workload: 7B 4-bit Qwen2.5 decode on Metal

## What I measured
I added local instrumentation to Candle's Metal backend to log pooled / in-use buffer bytes at command-buffer completion, plus per-command-buffer GPU completion times.

During a single request:
- pooled bytes grew from roughly `5058 MB` to roughly `11199 MB`
- in-use bytes grew from roughly `5007 MB` to roughly `11195 MB`
- the request emitted `50` completed command buffers totaling about `24.7 s` of GPU time

At the same time, isolated kernel benchmarks for the hot decode projections were fast (sub-ms to a few ms), so the pathological behavior only appeared under queued real-model execution.

## Important detail
A simple cap experiment did **not** fix this:
- forcing `wait_until_completed()` + dropping unused buffers when the pool crossed a threshold made throughput worse
- the pool still remained above the cap because in-flight buffers could not be reclaimed

That suggests the issue is not just "cache too large after completion"; it is also about how long transient outputs stay resident while command buffers are still outstanding.

## Why I'm filing this
On large-memory Apple machines this may be tolerable, but on small unified-memory devices it can turn an otherwise-correct Metal workload into swap-bound execution.

The key gap seems to be that the current pool treats all Metal buffers similarly. For this kind of workload, there is a big difference between:
- persistent model weights that should stay resident
- transient intermediate outputs that should be reclaimed or avoided as aggressively as possible

## Questions
1. Is there already a supported way to mark certain Metal allocations as transient / non-pooled?
2. Are there known plans to separate persistent and transient buffers in the Metal backend?
3. Is there a recommended way to bound pool growth on small unified-memory devices without forcing global waits?

If useful, I can follow up with a reduced reproduction or the exact instrumentation patches I used to capture pool growth and command-buffer timings.

**Comment by cjchanh** [comment]
Following up with the concrete measurements behind this report. These were captured while integrating a fused 4-bit Metal path in Cake, but the behavior below is about Candle's Metal allocator / pooled buffer lifetime rather than the custom kernel itself.

## Measured evidence

**Setup**
- `candle-core = 0.9.2`
- `candle-metal-kernels = 0.9.2`
- hardware: MacBook Air M1, 8 GB unified memory
- workload: `mlx-community/Qwen2.5-7B-Instruct-4bit` decode on Metal

**Individual kernel benchmarks (isolated, warm steady-state)**
- `q4_matvec 3584 -> 4608`: `0.817 ms` (vs Candle F16 matmul `1.711 ms`)
- `q4_matvec 3584 -> 37888`: `1.751 ms` (vs Candle F16 `5.977 ms`)
- `q4_matvec 3584 -> 152064`: `5.310 ms` (vs Candle F16 `23.123 ms`)
- rough total budget for `196` decode dispatches: `~381 ms`

**Actual decode throughput**
- `0.04 tok/s` (`~25 s/token`)

**Buffer-pool growth during the same decode request**
- pool snapshot grew from `~5058 MB` to `~11199 MB`
- in-use snapshot grew from `~5007 MB` to `~11195 MB`
- device physical RAM: `8192 MB`
- macOS swap activity was active during the run (swapins / swapouts in the millions)

**Command-buffer GPU timestamps**
- `50` completed command buffers for the request
- total GPU time: `24,697 ms`
- slowest CBs: `5558 ms`, `5399 ms`, `1599 ms`, `1538 ms`, `1529 ms`
- these are GPU completion timestamps, not CPU-side enqueue times

## Experiments that ruled out other causes
- CPU dispatch overhead: `1-29 us` per kernel
- zero explicit `synchronize()` calls during `forward()`
- per-layer GPU fence: no improvement
- GPU argmax + scalar readback: no improvement
- pool-cap experiment (`wait_until_completed()` + drop unused buffers): made throughput worse by turning the run into many tiny churn-heavy command buffers

## Working conclusion
These measurements point to queued transient Metal outputs accumulating faster than they can be reclaimed on small unified-memory devices. Once the pooled / in-use working set moves well past physical RAM, the workload appears to become swap-bound, and command buffers that should be millisecond-scale stretch into multi-second GPU runtimes under real execution.

The same model and fused q4 path on an M5 Max (128 GB unified memory) runs at `56.71 tok/s`, which is why this looks like a memory-pressure / buffer-lifetime constraint on small unified-memory machines rather than a raw kernel-throughput problem.

If useful, I can follow up with the exact instrumentation patches used to capture:
- per-command-buffer GPU start/end timestamps
- pooled vs in-use Metal buffer bytes
- the failed pool-cap experiment

**Comment by nuri-yoo** [comment]
I instrumented `MetalDevice::drop_unused_buffers` to log pool stats and ran `quantized-qwen2-instruct --which 7b` on an M4 Pro (36 GB). Here's what I see during decode:

```
[metal-pool] shared: 6311.6 MB (342 bufs, 342 in-use) | private: 22.0 MB (76 bufs, 69 in-use) | total: 6333.6 MB
```

The pattern is stable across all decode steps — shared pool never shrinks because all 342 buffers are permanently referenced by model weights.

**Root causes from code inspection:**

1. **`new_buffer_with_data` pools permanent buffers** — Model weights loaded via `new_buffer_with_data` are added to `self.buffers`, but they're held by the model struct for the entire lifetime. `drop_unused_buffers` can never reclaim them (`strong_count` is always > 1), and `find_available_buffer` scans all 342 entries every allocation for no benefit.

2. **`drop_unused_buffers` skips the private pool** — It only sweeps `self.buffers` (shared). Transient compute outputs in `self.private_buffers` are never cleaned, causing the slow growth I measured (~0.1-0.2 MB/token).

On an 8 GB M1, the 6.3 GB shared pool alone exceeds physical RAM, which matches the swap thrash you observed.

**Comment by laurigates** [comment]
Another data point for the allocator-lifetime theory, from a different manifestation: a **hard buffer-creation failure with plenty of physical headroom**, rather than growth into swap.

Environment: `candle-core 0.10.2` (as burn 0.21's `burn-candle` backend), Apple M4 Pro, **48 GB** unified memory, macOS 26.5.1. Workload: one LoRA training step over a bf16 ~12B-param diffusion transformer (Krea-2-Raw, ~25 GB of weights).

1. Loading the checkpoint straight onto the Metal device failed partway through the ~600-tensor load with `Metal error Failed to create metal resource: Buffer`.
2. Workaround — load on CPU, then move tensors to the device — got the full ~25 GB resident.
3. The first forward matmul then died the same way, despite the activations being MB-scale and ~20 GB of nominal headroom:

```
panicked at burn-candle-0.21.0/src/ops/tensor.rs:157:
called `Result::unwrap()` on an `Err` value: Metal error Failed to create metal resource: Buffer
```

(that line is `lhs.broadcast_matmul(&rhs).unwrap()`.)

At small scale the same stack is numerically sound for us (gradients within bf16 tolerance of a CPU reference), so this looks purely like buffer lifetime/pool behavior — possibly transient buffers pushing the total past the device working-set limit while the pool holds them, matching the "in-flight buffers can't be reclaimed" observation above. The failure is deterministic here. The repro is heavyweight (26 GB public model), but I'm happy to run instrumented builds on this 48 GB box if measurements from a larger-memory machine would help.
